File: src/email_parser/api/schemas/statistics.py
"""Statistics schemas for API."""

from datetime import date

from pydantic import BaseModel

from email_parser.models.application import ApplicationStatus, JobBoard


class SummaryStats(BaseModel):
    """Overall application statistics."""

    total_applications: int
    total_applicants: int
    status_counts: dict[str, int]
    source_counts: dict[str, int]
    review_rate: float = 0.0
    contact_rate: float = 0.0
    rejected_rate: float = 0.0
    hire_rate: float = 0.0


class MultiPositionApplicant(BaseModel):
    """Applicant who applied to more than one position."""

    applicant_name: str
    position_titles: str  # comma-separated
    application_link: str | None


class MultiPositionApplicantsResponse(BaseModel):
    """Response for applicants who applied to multiple positions."""

    total: int
    items: list[MultiPositionApplicant]


class PositionBreakdown(BaseModel):
    """Breakdown of applications by position."""

    position_title: str
    count: int
    percentage: float


class DailyCount(BaseModel):
    """Daily application count."""

    date: date
    count: int


class SourceBreakdown(BaseModel):
    """Breakdown of applications by source."""

    source: str
    count: int
    percentage: float


class StatusBySource(BaseModel):
    """Status distribution per source."""

    source: str
    status_counts: dict[str, int]


class DailyCountBySource(BaseModel):
    """Daily application count broken down by source."""

    date: date
    counts: dict[str, int]

File: frontend/src/api/statistics.ts
import client from './client';
import {
  DailyCount,
  DailyCountBySource,
  MultiPositionApplicantsResponse,
  PositionBreakdown,
  SourceBreakdown,
  StatusBySource,
  SummaryStats,
} from '../types';

export async function fetchSummaryStats(): Promise<SummaryStats> {
  const response = await client.get<SummaryStats>('/statistics/summary');
  return response.data;
}

export async function fetchDailyCounts(days: number = 30): Promise<DailyCount[]> {
  const response = await client.get<DailyCount[]>(`/statistics/daily?days=${days}`);
  return response.data;
}

export async function fetchDailyCountsBySource(days: number = 30): Promise<DailyCountBySource[]> {
  const response = await client.get<DailyCountBySource[]>(`/statistics/daily-by-source?days=${days}`);
  return response.data;
}

export async function fetchSourceBreakdown(): Promise<SourceBreakdown[]> {
  const response = await client.get<SourceBreakdown[]>('/statistics/by-source');
  return response.data;
}

export async function fetchStatusBySource(): Promise<StatusBySource[]> {
  const response = await client.get<StatusBySource[]>('/statistics/status-by-source');
  return response.data;
}

export async function fetchPositionBreakdown(): Promise<PositionBreakdown[]> {
  const response = await client.get<PositionBreakdown[]>('/statistics/by-position');
  return response.data;
}

export async function fetchAggregatedPositionBreakdown(): Promise<PositionBreakdown[]> {
  const response = await client.get<PositionBreakdown[]>(
    '/statistics/by-position-aggregated'
  );
  return response.data;
}

export async function fetchMultiPositionApplicants(): Promise<MultiPositionApplicantsResponse> {
  const response = await client.get<MultiPositionApplicantsResponse>(
    '/statistics/multi-position-applicants'
  );
  return response.data;
}
